### ISSUE TITLE
Recording Engine Problems

### ISSUE THREAD

**Issue by digipigeon** [sipwise/rtpengine#721]
Hello,

I have the following problem with the recording daemon (Version: 7.2.0.0+0~mr7.2.0.0 git-master-497c2d6)

If I set `output-mixed=1` I get a 78 byte -mix.wav file create and the log file shows

```
WARNING: [C yV9FKIhnAjXCYSeoFoBhKQ..-b95973b70d1055ae.meta] [S tag-0-media-1-component-1-RTP-id-2] [0x506fac16] Unknown RTP payload type 95
```

However if I set `output-single = 1`

I don't get any files created, but in the log files I get

```
Mar  2 14:37:25 rtpengine-recording[18797]: [1551537445.763084] ERR: [C 6zTKmaTilWGFmeizt-2LcA..-e789d489b8ef7fd2.meta] [S tag-1-media-1-component-1-RTP-id-0] [0x1bbd06f8] Failed to decode media packet
Mar  2 14:37:25 rtpengine-recording[18797]: [1551537445.773090] ERR: [C 6zTKmaTilWGFmeizt-2LcA..-e789d489b8ef7fd2.meta] [S tag-0-media-1-component-1-RTP-id-2] [0x1c5b0c46] Error resampling: failed to get resample buffers (code -22)
```

I am using ZoIPer with only ulaw and alaw enabled. Call is outbound from ZoIPer and error start occurring immediately after the offer. (many lines)

Please advise, thank you.

Kind Regards Jonathan

**Comment by rfuchs** [comment]
Post a full debug log (from rtpengine itself) for such a call please.

**Comment by digipigeon** [comment]
Please see: [rtpengine-recording-95.txt](https://github.com/sipwise/rtpengine/files/2922371/rtpengine-recording-95.txt)

**Comment by rfuchs** [comment]
No idea. You'll have to pull a pcap to see what that is. But that shouldn't keep recording from working. Maybe a bad ffmpeg lib version? The log looks straight forward, there's no reason it shouldn't be working.

**Comment by rfuchs** [comment]
Ok, so in the pcap you've sent me, this is packets 90 and 91 (in and out respectively). It's an RTP header with a single zero byte as payload, and unspecified payload type 95. A quick googling brought up this: https://community.zoiper.com/2730/what-is-payload-type-95

So this is actually fine and is (aside from the log message) ignored by both rtpengine and the recording daemon.

I took your pcap and played the RTP packets (including the PT 95 one) back into my locally running recording daemon. It processed everything fine and produced playable output files. So it's very likely that this is in fact a problem with your ffmpeg libs. Which ones (and which version) are you using?

**Comment by digipigeon** [comment]
Hi Richard,

Thank you for your time investigating this. I am using the included main repo libs from Ubuntu 18.04.2 LTS, these are primarily ```libavcodec-extra``` and ```libavfilter-dev``` installed with ```apt-get```. I think from aptitude this is 3.4.4

(```ffmpeg -version``` gives me no output as it does not appear to be in full.)

Do you know of any particular problems with 3.4.4? is there any recommended min version I should use?

Regards Jonathan

**Comment by rfuchs** [comment]
Cannot confirm or reproduce this on Ubuntu 18.04. It decodes just fine and produces audio files.

```
root@dfx-VirtualBox:~# dpkg -l libav\*|cat
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                       Version                  Architecture Description
+++-==========================-========================-============-=============================================================================
...
ii  libavcodec-dev:amd64       7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library with de/encoders for audio/video codecs - development files
un  libavcodec-extra57         <none>                   <none>       (no description available)
ii  libavcodec57:amd64         7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library with de/encoders for audio/video codecs - runtime files
ii  libavfilter-dev:amd64      7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library containing media filters - development files
un  libavfilter-extra6         <none>                   <none>       (no description available)
ii  libavfilter6:amd64         7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library containing media filters - runtime files
ii  libavformat-dev:amd64      7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library with (de)muxers for multimedia containers - development files
ii  libavformat57:amd64        7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library with (de)muxers for multimedia containers - runtime files
ii  libavresample-dev:amd64    7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg compatibility library for resampling - development files
ii  libavresample3:amd64       7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg compatibility library for resampling - runtime files
ii  libavutil-dev:amd64        7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library with functions for simplifying programming - development files
ii  libavutil55:amd64          7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library with functions for simplifying programming - runtime files
root@dfx-VirtualBox:~# dpkg -l libsw\*|cat
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                    Version                  Architecture Description
+++-=======================-========================-============-============================================================================
ii  libswresample-dev:amd64 7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library for audio resampling, rematrixing etc. - development files
ii  libswresample2:amd64    7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library for audio resampling, rematrixing etc. - runtime files
ii  libswscale-dev:amd64    7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library for image scaling and various conversions - development files
ii  libswscale4:amd64       7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library for image scaling and various conversions - runtime files
```

**Comment by digipigeon** [comment]
Just FYI here are my outputs for the same

```
# dpkg -l libav\*|cat
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                     Version                  Architecture Description
+++-========================-========================-============-=============================================================================
ii  libavcodec-dev:amd64     7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library with de/encoders for audio/video codecs - development files
ii  libavcodec-extra         7:3.4.4-0ubuntu0.18.04.1 all          FFmpeg library with extra codecs (metapackage)
ii  libavcodec-extra57:amd64 7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library with additional de/encoders for audio/video codecs
un  libavcodec57             <none>                   <none>       (no description available)
ii  libavfilter-dev:amd64    7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library containing media filters - development files
un  libavfilter-extra6       <none>                   <none>       (no description available)
ii  libavfilter6:amd64       7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library containing media filters - runtime files
ii  libavformat-dev:amd64    7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library with (de)muxers for multimedia containers - development files
ii  libavformat57:amd64      7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library with (de)muxers for multimedia containers - runtime files
ii  libavresample-dev:amd64  7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg compatibility library for resampling - development files
ii  libavresample3:amd64     7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg compatibility library for resampling - runtime files
ii  libavutil-dev:amd64      7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library with functions for simplifying programming - development files
ii  libavutil55:amd64        7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library with functions for simplifying programming - runtime files
# dpkg -l libsw\*|cat
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                    Version                  Architecture Description
+++-=======================-========================-============-============================================================================
ii  libswresample-dev:amd64 7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library for audio resampling, rematrixing etc. - development files
ii  libswresample2:amd64    7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library for audio resampling, rematrixing etc. - runtime files
ii  libswscale-dev:amd64    7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library for image scaling and various conversions - development files
ii  libswscale4:amd64       7:3.4.4-0ubuntu0.18.04.1 amd64        FFmpeg library for image scaling and various conversions - runtime files
```

Please let me know if you notice anything amiss. Otherwise do you have any ideas what else I could try to debug this problem further?

**Comment by rfuchs** [comment]
What's the full config for the recording daemon you're using?

**Comment by digipigeon** [comment]
```
[rtpengine-recording]

table = 0

### number of worker threads (default 8)
# num-threads = 16

### where to forward to (unix socket)
# forward-to = /var/run/rtpengine/sock

### where to store recordings: file (default), db, both
# output-storage = db

### format of stored recordings: wav (default), mp3
# output-format = mp3

### directory containing rtpengine metadata files
spool-dir = /var/spool/rtpengine

### where to store media files to
output-dir = /var/spool/rtpengine-recording

### resample all output audio
# resample-to = 8000

### bits per second for MP3 encoding
# mp3_bitrate = 24000

### mix participating sources into a single output
output-mixed = 1

### create one output file for each source
# output-single = 1

### mysql configuration for db storage
#mysql-host = [secret]
#mysql-port = 3306
#mysql-user = rtpengine
#mysql-pass = [secret]
#mysql-db = rtpengine
```

**Comment by rfuchs** [comment]
So that's the same I was using. I've created branch `rfuchs/gh721` with some extra logging messages, in particular 7616417 and cd53278

**Comment by digipigeon** [comment]
Please find attached log file for the version compiled from rfuchs/gh721

[rtpengine-recording-95-2.txt](https://github.com/sipwise/rtpengine/files/2956594/rtpengine-recording-95-2.txt)

Please let me know if you find anything useful in here.

Thank you, Regards Jonathan

**Comment by rfuchs** [comment]
I would need the log from the recording daemon

**Comment by digipigeon** [comment]
This should be in here, such as this line
```
Mar 12 11:22:45 rtp-test rtpengine-recording[4932]: [1552389765.180535] WARNING: [C qxc3mMqIWDbZCTAelmFGlg..-1ca35dbdf90c9bb6.meta] [S tag-0-media-1-component-1-RTP-id-2] [0xf6ef6be2] Unknown RTP payload type 95
```

Are there any options to increase verbosity of recording daemon?

**Comment by digipigeon** [comment]
[recording-error.txt](https://github.com/sipwise/rtpengine/files/3053568/recording-error.txt)

Please see the attached file, which has your custom log information present.

**Comment by rfuchs** [comment]
Ok, so the output format is left at -1, which is wrong. Can you look in the same log file, somewhere around where the recording was started, for a line that reads `using output sample format ...`, and lines that read `supported sample format for output codec ...` just above it? And post those?

**Comment by digipigeon** [comment]
[recording-error2.txt](https://github.com/sipwise/rtpengine/files/3054646/recording-error2.txt)
I was not able to see these exact lines, but please see the attached file containing the full logs (new call).

**Comment by rfuchs** [comment]
That doesn't include the debug log messages as it did before.

**Comment by digipigeon** [comment]
My Apologies, please see this file:
[recording-error3.txt](https://github.com/sipwise/rtpengine/files/3054728/recording-error3.txt)

**Comment by rfuchs** [comment]
Any idea why there's two different PIDs appearing in this log for the recording daemon? 2191 and 553?

**Comment by digipigeon** [comment]
I am guessing that may be due to me having difficulty getting the logging to work correctly. If this has caused problems identifying the problem I can try again.

**Comment by rfuchs** [comment]
No I'm just wondering if there's maybe two instances running? Or if those are duplicate log messages?

**Comment by digipigeon** [comment]
There is 2 instances running, this was my mistake, one backgrounded from the shell, the other as a service.

**Comment by rfuchs** [comment]
Ok I see what's missing, the encoder config is somehow never done. But from the code I don't see how that could be happening. I've updated the `gh721` branch with yet more debug logging for this, see commit 44a7c41

**Comment by digipigeon** [comment]
[recording-error4.txt](https://github.com/sipwise/rtpengine/files/3055489/recording-error4.txt)
Please see attached.

**Comment by rfuchs** [comment]
That seems to be from the previous build. It doesn't have any of the additional logging messages. Also the second instance with PID 553 is still running.

**Comment by digipigeon** [comment]
I am sorry for the mess in my provided files, I will get a clean log back for you tomorrow morning.

**Comment by digipigeon** [comment]
[recording-error5.txt](https://github.com/sipwise/rtpengine/files/3055592/recording-error5.txt)
Please find attached.

**Comment by rfuchs** [comment]
Does it make a difference if you use `output-mixed = true` instead of `output-mixed = 1` ?

**Comment by rfuchs** [comment]
Also, does it make a difference if in your signalling you don't use `record-call=yes` but rather just `record-call` ?

**Comment by juha-h** [comment]
Jonathan Hulme writes:

> Setting ```record-call``` instead of ```record-call=yes``` seems to
> have fixed this problem.

I have used 'record-call=on' and it has worked.  What is the current
way?

**Comment by rfuchs** [comment]
They should both work. I have a bugfix ready, just wanted to confirm that this is actually the problem.

**Comment by rfuchs** [comment]
Fixed in master and 7.2
